src/jit/codegenarm64.cpp

```diff
@@ -599,108 +599,116 @@ void CodeGen::genRestoreCalleeSavedRegistersHelp(regMaskTP   regsToRestoreMask,
+    unsigned intRegsToRestoreCount = genCountBits(maskRestoreRegsInt);
+    int stackDelta = 0;
+    bool isPairRestore = false;
+    bool thisIsTheLastRestoreInstruction = false;
+#ifdef DEBUG
+    bool isRegsToRestoreCountOdd = ((floatRegsToRestoreCount + intRegsToRestoreCount) % 2 != 0);
+#endif
+
     // We want to restore in the opposite order we saved, so the unwind codes match. Be careful to handle odd numbers of
     // callee-saved registers properly.
 
-    if (maskRestoreRegsFloat != RBM_NONE)
-    {
-        // Restore the floating-point/SIMD registers
-
-        unsigned floatRegsToRestoreCount = genCountBits(maskRestoreRegsFloat);
-
-        while (maskRestoreRegsFloat != RBM_NONE)
-        {
-            if ((floatRegsToRestoreCount % 2) == 0)
-            {
-                assert(floatRegsToRestoreCount >= 2);
-
-                regMaskTP reg2Mask = genFindHighestBit(maskRestoreRegsFloat);
-                regNumber reg2 = genRegNumFromMask(reg2Mask);
-                maskRestoreRegsFloat &= ~reg2Mask;
-
-                regMaskTP reg1Mask = genFindHighestBit(maskRestoreRegsFloat);
-                regNumber reg1 = genRegNumFromMask(reg1Mask);
-                maskRestoreRegsFloat &= ~reg1Mask;
-
-                spOffset -= 2 * FPSAVE_REGSIZE_BYTES;
-
-                // Is this the last restore instruction? And have we've been told to adjust SP?
-                bool thisIsTheLastRestoreInstruction = (floatRegsToRestoreCount == 2) && (maskRestoreRegsInt == RBM_NONE);
-                genEpilogRestoreRegPair(reg1, reg2, spOffset, thisIsTheLastRestoreInstruction ? spDelta : 0, REG_IP0, nullptr);
 
-                floatRegsToRestoreCount -= 2;
-            }
-            else
-            {
-                // We do the odd register first when restoring, last when saving.
-                assert((floatRegsToRestoreCount % 2) == 1);
+    // Restore the floating-point/SIMD registers
 
-                regMaskTP reg1Mask = genFindHighestBit(maskRestoreRegsFloat);
-                regNumber reg1 = genRegNumFromMask(reg1Mask);
-                maskRestoreRegsFloat &= ~reg1Mask;
-
-                spOffset -= FPSAVE_REGSIZE_BYTES;
+    while (maskRestoreRegsFloat != RBM_NONE)
+    {
+        thisIsTheLastRestoreInstruction = (floatRegsToRestoreCount <= 2) && (maskRestoreRegsInt == RBM_NONE);
+        isPairRestore = (floatRegsToRestoreCount % 2) == 0;
 
-                // Is this the last restore instruction? And have we've been told to adjust SP?
-                bool thisIsTheLastRestoreInstruction = (floatRegsToRestoreCount == 1) && (maskRestoreRegsInt == RBM_NONE);
-                genEpilogRestoreReg(reg1, spOffset, thisIsTheLastRestoreInstruction ? spDelta : 0, REG_IP0, nullptr);
+        // Update stack delta only if it is the last restore (the first save).
+        if (thisIsTheLastRestoreInstruction)
+        {
+            assert(stackDelta == 0);
+            stackDelta = spDelta;
+        }
 
```

Comment:
It's more about safety that we set this only once (from 0 to some value) throughout INT/FP registers.